src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -65,156 +66,57 @@ public class TemplateBinder
-                var hasNewParameterValue = values.TryGetValue(parameterName, out newParameterValue);
-
-                object currentParameterValue = null;
-                var hasCurrentParameterValue = ambientValues != null &&
-                                               ambientValues.TryGetValue(parameterName, out currentParameterValue);
+            return GetValues(ambientValues: ambientValues, explicitValues: values, requiredValuesKeys: null);
+        }
 
-                if (hasNewParameterValue && hasCurrentParameterValue)
-                {
-                    if (!RoutePartsEqual(currentParameterValue, newParameterValue))
-                    {
-                        // Stop copying current values when we find one that doesn't match
-                        break;
-                    }
-                }
+        internal TemplateValuesResult GetValues(
+            RouteValueDictionary ambientValues,
+            RouteValueDictionary explicitValues,
+            IEnumerable<string> requiredValuesKeys)
+        {
+            var context = new TemplateBindingContext(_defaults);
 
-                if (!hasNewParameterValue &&
-                    !hasCurrentParameterValue &&
-                    _defaults?.ContainsKey(parameter.Name) != true)
-                {
-                    // This is an unsatisfied parameter value and there are no defaults. We might still
-                    // be able to generate a URL but we should stop 'accepting' ambient values.
-                    //
-                    // This might be a case like:
-                    //  template: a/{b?}/{c?}
-                    //  ambient: { c = 17 }
-                    //  values: { }
-                    //
-                    // We can still generate a URL from this ("/a") but we shouldn't accept 'c' because
-                    // we can't use it.
-                    // 
-                    // In the example above we should fall into this block for 'b'.
-                    break;
-                }
+            var canCopyParameterAmbientValues = true;
 
-                // If the parameter is a match, add it to the list of values we will use for URI generation
-                if (hasNewParameterValue)
-                {
-                    if (IsRoutePartNonEmpty(newParameterValue))
-                    {
-                        context.Accept(parameterName, newParameterValue);
-                    }
-                }
-                else
-                {
-                    if (hasCurrentParameterValue)
-                    {
-                        context.Accept(parameterName, currentParameterValue);
-                    }
-                }
+            // In case a route template is not parameterized, we do not what the required parameters are, so look
+            // at required values of an endpoint to get the key names and then use them to decided if ambient values
+            // should be used.
```

Comment:
It would really help to include an example in the comments of what cases like this would look like. Feel free to add comments explaining how MVC uses it.